deps/triemap/triemap.c

```diff
@@ -540,60 +541,49 @@ void TrieMapNode_Free(TrieMapNode *n, void (*freeCB)(void *)) {
 }
 
 /* the current top of the iterator stack */
-#define __tmi_current(it) &it->stack[it->stackOffset - 1]
+#define __tmi_current(it) &array_tail(it->stack)
 
 /* Step itearator return codes below: */
 
 #define TM_ITERSTATE_SELF 0
 #define TM_ITERSTATE_CHILDREN 1
 
 /* Push a new trie node on the iterator's stack */
-inline void __tmi_Push(TrieMapIterator *it, TrieMapNode *node) {
-  if (it->stackOffset == it->stackCap) {
-    // make sure we don't overflow stackCap
-    if ((int)it->stackCap + 1024 > 0xFFFF) {
-      it->stackCap = 0xFFFF;
-    } else {
-      it->stackCap += MIN(it->stackCap, 1024);
-    }
-    it->stack = rm_realloc(it->stack, it->stackCap * sizeof(__tmi_stackNode));
-  }
-  it->stack[it->stackOffset++] = (__tmi_stackNode){
+inline void __tmi_Push(TrieMapIterator *it, TrieMapNode *node, tm_len_t stringOffset,
+                       bool found) {
+  __tmi_stackNode stackNode = {
       .childOffset = 0,
-      .stringOffset = 0,
+      .stringOffset = stringOffset,
+      .found = found,
       .n = node,
       .state = TM_ITERSTATE_SELF,
   };
+  it->stack = array_ensure_append_1(it->stack, stackNode);
 }
 
 inline void __tmi_Pop(TrieMapIterator *it) {
-  it->bufOffset -= (__tmi_current(it))->stringOffset;
-  if (it->bufOffset < it->prefixLen) {
-    it->inSuffix = 0;
-  }
-  --it->stackOffset;
+  __tmi_stackNode *current = __tmi_current(it);
+  it->buf = array_trimm_len(it->buf, current->stringOffset);
+  array_pop(it->stack);
 }
 
 TrieMapIterator *TrieMap_Iterate(TrieMap *t, const char *prefix, tm_len_t len) {
   TrieMapIterator *it = rm_calloc(1, sizeof(TrieMapIterator));
 
-  it->bufLen = 16;
-  it->buf = rm_calloc(1, it->bufLen);
-  it->stackCap = 8;
-  it->stack = rm_calloc(it->stackCap, sizeof(__tmi_stackNode));
-  it->bufOffset = 0;
-  it->inSuffix = 0;
+  it->buf = array_new(char, 16);
+  it->stack = array_new(__tmi_stackNode, 8);
   it->prefix = prefix;
   it->prefixLen = len;
+  it->mode = TM_PREFIX_MODE;
```

Comment:
Can you add a comment here what this (new?) mode field?

Reply:
Added comment in code.
The iterator used to be only prefix mode. Now it has more modes.